test/common/tls/integration/ssl_integration_test.cc

```diff
@@ -1164,5 +1177,107 @@ TEST_P(SslKeyLogTest, SetMultipleIps) {
   logCheck();
 }
 
+TEST_P(SslIntegrationTest, SyncCertSelectorSucceeds) {
+  tls_cert_selector_yaml_ = R"EOF(
+name: test-tls-context-provider
+typed_config:
+  "@type": type.googleapis.com/google.protobuf.StringValue
+  value: sync
+  )EOF";
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection({});
+  };
+  testRouterRequestAndResponseWithBody(16 * 1024 * 1024, 16 * 1024 * 1024, false, false, &creator);
+  checkStats();
+}
+
+TEST_P(SslIntegrationTest, AsyncCertSelectorSucceeds) {
+  tls_cert_selector_yaml_ = R"EOF(
+name: test-tls-context-provider
+typed_config:
+  "@type": type.googleapis.com/google.protobuf.StringValue
+  value: async
+  )EOF";
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection({});
+  };
+  testRouterRequestAndResponseWithBody(16 * 1024 * 1024, 16 * 1024 * 1024, false, false, &creator);
+  checkStats();
+}
+
+TEST_P(SslIntegrationTest, AsyncSleepCertSelectorSucceeds) {
+  tls_cert_selector_yaml_ = R"EOF(
+name: test-tls-context-provider
+typed_config:
+  "@type": type.googleapis.com/google.protobuf.StringValue
+  value: sleep
+  )EOF";
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection({});
+  };
+  testRouterRequestAndResponseWithBody(16 * 1024 * 1024, 16 * 1024 * 1024, false, false, &creator);
+  checkStats();
+}
+
+TEST_P(SslIntegrationTest, AsyncSleepCertSelectionAfterTearDown) {
+  tls_cert_selector_yaml_ = R"EOF(
+name: test-tls-context-provider
+typed_config:
+  "@type": type.googleapis.com/google.protobuf.StringValue
+  value: sleep
+  )EOF";
+  initialize();
+
+  Network::ClientConnectionPtr connection = makeSslClientConnection({});
+  ConnectionStatusCallbacks callbacks;
+  connection->addConnectionCallbacks(callbacks);
+  connection->connect();
+  const auto* socket = dynamic_cast<const Extensions::TransportSockets::Tls::SslHandshakerImpl*>(
+      connection->ssl().get());
+  ASSERT(socket);
+  // wait for 200ms, let the server tls handshake into sleep state.
+  waitForTime(std::chrono::milliseconds(200));
```

Comment:
Instead of sleep, please wait for a metric or other event you can observe. You can add metrics to your test cert selector if needed.